File: templates/python/ribbon_3d.py
"""ribbon_3d: 多条 3D 条带（多曲线时间演化）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from theme import apply_theme
from palette import cycle

def make_figure(title='3D ribbons'):
    apply_theme(fig_size=(6.5, 5))
    x = np.linspace(0, 10, 80)
    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')
    width = 0.3
    for i in range(5):
        z = np.sin(x + i*0.6) * np.exp(-x/8)
        verts = []
        for j in range(len(x)-1):
            verts.append([
                (x[j],   i - width, z[j]),
                (x[j+1], i - width, z[j+1]),
                (x[j+1], i + width, z[j+1]),
                (x[j],   i + width, z[j]),
            ])
        poly = Poly3DCollection(verts, facecolor=cycle(i), edgecolor='none', alpha=0.85)
        ax.add_collection3d(poly)
    ax.set_xlim(0, 10); ax.set_ylim(-1, 5); ax.set_zlim(-1, 1.2)
    ax.set_xlabel('x'); ax.set_ylabel('series'); ax.set_zlabel('z')
    ax.set_title(title); ax.view_init(30, -60)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/ribbon_3d.m
function fig = ribbon_3d()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme();
    x = linspace(0, 10, 80)';
    Y = zeros(numel(x), 5);
    for i = 1:5
        Y(:, i) = sin(x + (i-1)*0.6) .* exp(-x/8);
    end
    fig = figure('Position',[100 100 650 500]);
    h = ribbon(x, Y);
    for k = 1:numel(h), h(k).FaceColor = palette('cat', k); h(k).EdgeColor = 'none'; end
    xlabel('series'); ylabel('x'); zlabel('value'); title('3D ribbons');
    view(45, 30);
end
